### ISSUE TITLE
Unable to push commits to the website after this week's updates

### ISSUE THREAD

**Issue by xt0rted** [desktop/desktop#1727]
### Description

After updating to 0.5.6 and 0.5.7 I'm no longer able to push my branch's commits. This works from the command line as well as the native client but no longer works in this client. I'm unsure if this was an issue with 0.5.5 but last week I was able to push my commits just fine.

### Version
**GitHub Desktop version:** 0.5.6, 0.5.7
**OS version:** Win10 Creators Update (10.0.15063)

### Steps to Reproduce

I'm not sure if this is reproducible elsewhere, but it keeps happening on the repo I'm working in even after updating from 0.5.6 to 0.5.7.

I'm pressing the button to push my changes to origin, as well as using the push menu. Both result in similar errors to this.

![image](https://cloud.githubusercontent.com/assets/831974/26377667/3a381142-3fdf-11e7-8440-69d752a9ea44.png)

**Expected behaviour:** My new commits should be pushed to GitHub

**Actual behaviour:** I get an error message and nothing is pushed

**Reproduces how often:** Every time I try to push for this repo

### Logs

There are no log files in `%APPDATA%\GitHub Desktop\logs\*.desktop.production.log`

#### Additional Information

It looks like the issue is being triggered by `-c credential.helper=` because when I include that on the command line I'm prompted for my username & password. According to the application I'm logged in.

**Comment by shiftkey** [comment]
@xt0rted could you confirm your account is still visible under Preferences?

**Comment by niik** [comment]
> It looks like the issue is being triggered by -c credential.helper= because when I include that on the command line I'm prompted for my username & password.

I'd expect that on the command line since we set a specific [GIT_ASKPASS environment variable](https://github.com/desktop/desktop/blob/3dc8738e711131047f9973398a5dac08e2cd206e/app/src/lib/git/core.ts#L230) to have Git talk to us instead of any configured credential helper. Seems to me like either something isn't cooperating for you inside of our askpass implementation or we're not providing the ask pass script with an account.

> According to the application I'm logged in.

Gotcha, and this is running the production build, right? I.e. you're not compiling and running this your self?

Is this just for pushes? Does fetch/pull still work?

Could you verify that the repository icon in the toolbar is a lock, ie like this

![image](https://cloud.githubusercontent.com/assets/634063/26411108/56c425b4-40a5-11e7-9cc6-6737c3ea91b3.png)

**Comment by xt0rted** [comment]
My account still shows under options (I don't see any Preferences so I'm guessing that's what was being referred to) ~~but when I try to switch to the Git or Advanced tabs the popup immediately closes. I'm not seeing anything logged in the console when this happens though.~~ Looks like that was fixed in todays update.

![image](https://cloud.githubusercontent.com/assets/831974/26413187/2b8bc5b8-4079-11e7-9e48-c10d159323e6.png)

All of my private repos that showed a lock icon are switching back to the book icon once they're selected. This is what shows in my log when that happens:
```log
2017-05-24T03:25:34.827Z - info: Stats reported.
2017-05-24T15:33:27.626Z - info: Opt in reported.
2017-05-24T15:33:32.225Z - info: Opt in reported.
2017-05-24T15:33:35.935Z - info: Opt in reported.
2017-05-24T15:33:49.507Z - info: Opt in reported.
2017-05-24T15:33:58.195Z - info: Opt in reported.
2017-05-24T15:34:24.777Z - info: Opt in reported.
2017-05-24T15:34:36.653Z - error: fetchRepository: not found for 'xt0rted' and 'equineexchange/email-templates'
Error: {
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3"
}

    at file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:77900
    at XMLHttpRequest.J.ae.onreadystatechange (file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:74777)
2017-05-24T15:34:40.549Z - error: fetchRepository: not found for 'xt0rted' and 'equineexchange/hubot'
Error: {
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3"
}

    at file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:77900
    at XMLHttpRequest.J.ae.onreadystatechange (file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:74777)
2017-05-24T15:34:45.574Z - error: fetchRepository: not found for 'xt0rted' and 'xt0rted/EEI'
Error: {
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3"
}

    at file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:77900
    at XMLHttpRequest.J.ae.onreadystatechange (file:///C:/Users/brian/AppData/Local/GitHubDesktop/app-0.5.7/resources/app/renderer.js:3:74777)
```

The repos in question are all forks, one of my own private repos doesn't seem to give any errors when clicking the Fetch origin button and its icon remains as a lock when I select it.

I did run a custom build while working on a PR last week, but I'm not running that and only did while testing the update.

**Comment by xt0rted** [comment]
I just tried to clone one of my private repos and that worked ok, but when I tried to clone one of the private repos from the EquineExchange org where all of the other repos are that are giving me issue I got an error that said:

> We couldn't find that repository. Check that you are logged in, and the URL or repository alias are spelled correctly.

My access to the org is ok because I can sync & clone with the native client as well as browse everything on the site.

**Comment by niik** [comment]
Thank you so much for the extra details @xt0rted! I'm looking into this now, I've found a few things that's certainly wrong with how we associate repositories at the moment but unfortunately they don't seem to be wrong enough for me to be able to reproduce this.

> I just tried to clone one of my private repos and that worked ok, but when I tried to clone one of the private repos from the EquineExchange org

How did you clone these? Through clicking on the clone button on the website or by entering the url manually in the client?

Also, could you please run the following command in one of the repositories that's causing these issues?

`git remote -v`

Finally, could you open devtools and look at the network tab as these requests are being made? Look for requests that contain your org name and take a screenshot of them. Be careful not to include the request data itself as that will (probably) include your private token.

**Comment by xt0rted** [comment]
> How did you clone these? Through clicking on the clone button on the website or by entering the url manually in the client?

I clicked the button on the website.

> Also, could you please run the following command in one of the repositories that's causing these issues?

This is the repo/branch I first ran into the issue with.

```powershell
C:\dev\equineexchange\EEI [issue-728 ≡ +0 ~15 -0 !]> git remote -v
equineexchange  https://github.com/equineexchange/EEI.git (fetch)
equineexchange  https://github.com/equineexchange/EEI.git (push)
origin  https://github.com/xt0rted/EEI.git (fetch)
origin  https://github.com/xt0rted/EEI.git (push)
upstream        https://github.com/equineexchange/EEI.git (fetch)
upstream        https://github.com/equineexchange/EEI.git (push)
```

The network stats are kind of interesting.

When I click the Fetch origin button nothing shows in the network stats but I still get the following error popup three times in a row:

![image](https://cloud.githubusercontent.com/assets/831974/26415253/d9d58e8c-407f-11e7-9ff4-fc941facbff0.png)

When I switch to the repo in question I get the following showing up:

![image](https://cloud.githubusercontent.com/assets/831974/26415169/93d7088e-407f-11e7-839d-7520ee6572d9.png)

It looks like the only time something shows in the network stats for this repo is when switching back and forth from it.

**Comment by xt0rted** [comment]
The repo I was testing with is a fork, this one is a clone from the org.

```powershell
C:\dev\equineexchange\design [master ≡ +13 ~2 -9 !]> git remote -v
origin  https://github.com/equineexchange/design.git (fetch)
origin  https://github.com/equineexchange/design.git (push)
```

![image](https://cloud.githubusercontent.com/assets/831974/26415744/8c972b4c-4081-11e7-8632-ddf545012878.png)

This is one of my private repos:

```powershell
C:\dev\GitHub\roxie-rose-design [master ≡ +1 ~0 -0 !]> git remote -v
origin  https://github.com/xt0rted/roxie-rose-design.git (fetch)
origin  https://github.com/xt0rted/roxie-rose-design.git (push)
```

![image](https://cloud.githubusercontent.com/assets/831974/26415810/c298e334-4081-11e7-90bc-01345fd8a691.png)

**Comment by joshaber** [comment]
@xt0rted Thanks for the great detective work!

It looks like the `equineexchange` org uses the restrictive application access policy (which is good!). Right now my best guess is this relates to that somehow. I'm going to do more testing tomorrow.

**Comment by xt0rted** [comment]
Should I try signing out and back in with the app or do we want to try and track down what's going on before doing that? I'm pushing/pulling with the native app and command line so this not working isn't an issue.

**Comment by niik** [comment]
@xt0rted Digging through our server side logs here I can't find any request to the repositories you mention that were made by the production OAuth app. I can, however, find several matching entries for our dev OAuth app that correspond to the requests you posted about. This makes me suspect that the production build of our app is reusing the token from the dev app (the one you built locally).

A casual look at our [logic for retrieving the token from the windows credential store](https://github.com/desktop/desktop/blob/0897141f8f455c9eb76d8b8263073fdb8a0529c9/app/src/lib/auth.ts#L9-L11) seems to support this conclusion. We only construct the key under which the token is stored as `GitHub - https://api.github.com/niik`for both the dev build and the production build where we really should be using something like `GitHub Dev - ...` for the dev build. I haven't been able to confirm this yet but it seems highly plausible to me.

This would explain why you're able to pull from your forks since, as @joshaber described, the org is blocking the dev build oauth app and allowing the production app.

> Should I try signing out and back in with the app or do we want to try and track down what's going on before doing that?

I'm fairly certain that what what I described above is the problem and signing out and back in with the production app should confirm that so please go ahead and do that. If it still doesn't work please take to DevTools again and copy the `X-GitHub-Request-Id` header of a failing request so that I can find the corresponding entry in our log files.

**Comment by niik** [comment]
Tentative fix for this in #1745 / 1fb5b3c155f7695057cc778229f12661551699d8

**Comment by xt0rted** [comment]
Signing out and back in fixed it.
